### PR TITLE
 Add docs about memory format

### PR DESCRIPTION
**Pull request by VitalyFedyunin**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#34818 [WIP] Add docs about memory format**

Differential Revision: [D20601336](https://our.internmc.facebook.com/intern/diff/D20601336)

### CODE REVIEW

**Review by jlin27 — APPROVED** [review]

File: torch/_tensor_docs.py [hunk 1]
```diff
@@ -739,15 +744,22 @@ def scale_channels(input, scale):
...
 
 add_docstr_all('contiguous',
                r"""
-contiguous() -> Tensor
+contiguous(memory_format=torch.contiguous_format) -> Tensor
 
-Returns a contiguous tensor containing the same data as :attr:`self` tensor. If
-:attr:`self` tensor is contiguous, this function returns the :attr:`self`
-tensor.
+Returns a contiguous in memory tensor containing the same data as :attr:`self` tensor. If
```
Comment (gchanan) [inline torch/_tensor_docs.py]: nit: contiguous-in-memory

File: torch/_tensor_docs.py [hunk 2]
```diff
@@ -739,15 +744,22 @@ def scale_channels(input, scale):
...
+:attr:`self` tensor is already in the specified memory format, this function returns the 
+:attr:`self` tensor.
+
+Args: 
+    memory_format (:class:`torch.memory_format`, optional): the desired memory format of
```
Comment (gchanan) [inline torch/_tensor_docs.py]: is preserve allowed?  Maybe worth mentioning it if it's not.

File: torch/_torch_docs.py [hunk 3]
```diff
@@ -84,6 +86,8 @@ def merge_dicts(*dicts):
         returned tensor. Default: ``False``.
     pin_memory (bool, optional): If set, returned tensor would be allocated in
         the pinned memory. Works only for CPU tensors. Default: ``False``.
+    memory_format (:class:`torch.memory_format`, optional): the desired memory format of 
+        returned Tensor. Default: ``torch.preserve_format``.
```
Comment (gchanan) [inline torch/_torch_docs.py]: nit: the returned Tensor.

File: torch/_tensor_docs.py [hunk 4]
```diff
@@ -8,6 +8,11 @@
 def add_docstr_all(method, docstr):
     add_docstr(getattr(torch._C._TensorBase, method), docstr)
 
+common_args = parse_kwargs("""
+    memory_format (:class:`torch.memory_format`, optional): the desired memory format of 
+        returned Tensor. Default: ``torch.preserve_format``.
```
Comment (gchanan) [inline torch/_tensor_docs.py]: nit: the returned Tensor.

File: torch/_tensor_docs.py [hunk 5]
```diff
@@ -739,15 +744,22 @@ def scale_channels(input, scale):
...
-Returns a contiguous tensor containing the same data as :attr:`self` tensor. If
-:attr:`self` tensor is contiguous, this function returns the :attr:`self`
-tensor.
+Returns a contiguous in memory tensor containing the same data as :attr:`self` tensor. If
+:attr:`self` tensor is already in the specified memory format, this function returns the 
+:attr:`self` tensor.
+
+Args: 
+    memory_format (:class:`torch.memory_format`, optional): the desired memory format of 
+        returned Tensor. Default: ``torch.contiguous_format``.
```
Comment (gchanan) [inline torch/_tensor_docs.py]: nit: the returned Tensor.

File: docs/source/tensor_attributes.rst [hunk 6]
```diff
@@ -210,3 +210,25 @@ Example::
     (1, 5)
 
 For more information on ``torch.sparse_coo`` tensors, see :ref:`sparse-docs`.
+
+torch.memory_format
+------------
+
+.. class:: torch.memory_format
+
+A :class:`torch.memory_format` is an object representing the memory format on which a :class:`torch.Tensor` is
```
Comment (gchanan) [inline docs/source/tensor_attributes.rst]: it's a bit non-idiomatic to structure things like "X is an object representing the X", it's more natural to say something like "X is an object representing 'longer description of X'".  Also remove the "A" at the beginning of the sentence.

Basically I'd expect some description about physical memory layout vs semantic dimension meaning here.

Something like (please improve!):

> `torch.memory_format` is an object that represents how the dimensions of a Tensor are physically laid out in memory.  Memory format allows users to separately control the physical ordering and semantic meaning of the dimensions of a Tensor.  For example, a user may wish to semantically represent a batch of images as having dimension order [N, C, H, W] (to match PyTorch's convention), but may wish to order the dimensions in memory as [N, H, W C] in order to take advantage of (possibly third-party) optimized algorithms for Tensors in that physical layout.

**Pull request closed without merging**